### ISSUE TITLE
Rewriter generates too many potential matches

### ISSUE THREAD

**Issue by brianhuffman** [GaloisInc/saw-script#1577]
While working on some saw proofs that prove a lot of small lemmas using rewriting, I noticed that many of the calls to the simplifier were taking more time than I thought they should. Profiling showed that the `apply` function (defined locally inside `rewriteSharedTerm`) was being called hundreds of times more often than the surrounding `rewriteTop` function, suggesting that `TermNet.unify_term` is returning a lot more potential matches than it should, and all these bogus potential matches need to be filtered out by calling the much more expensive function `scMatch`.

By adding some debug output to `rewriteSharedTerm`, I discovered that when rewriting with `cryptol_ss`, terms like `sort 0`, `Prelude.Nat`, or `Prelude.Vec 32 Prelude.Bool` yield a list of 161 potential rewrite rules. This is the smallest set of potential rewrites that arises from any term when simplified with `cryptol_ss`. The potential rules all seem to be ones that have saw-core record projections appearing on the left-hand side.

For other terms, including function types, product types, or natural numbers, the rewriter finds 430 potential rewrite rules. (I think this is actually the complete set of *all* the rules in `cryptol_ss`.)

Basically this means that the term net data structure is doing nothing for us. I believe the issue is in the `toPat` function that turns a saw-core `Term` into a term-net `Pat`: It looks like it's treating way too many term constructors as wildcards that could potentially match anything.

**Comment by brianhuffman** [comment]
The root of the problem is that this definition of `termToPat` has a catch-all case that maps lots of terms to the wildcard pattern `Var`. As new constructors have been added to the `Term` datatype, this function definition was not kept up to date.

https://github.com/GaloisInc/saw-script/blob/81287fb74cbdac3bb3f75b984f1125deb3055e64/saw-core/src/Verifier/SAW/Term/Functor.hs#L409-L420

I will rewrite this function so that it does not use the underscore catch-all case, instead using a complete pattern match. This way we will get a compiler warning if we forget to update `termToPat` after changing the `Term` datatype again.

**Comment by brianhuffman** [comment]
The other reason why we're returning way too many rewrite rules is that the rewriter queries the term net in "unify" mode instead of "match" mode: https://github.com/GaloisInc/saw-script/blob/81287fb74cbdac3bb3f75b984f1125deb3055e64/saw-core/src/Verifier/SAW/Rewriter.hs#L649

This change was introduced in 916040bbec1d46e3aad2901e179d76db85d63264 in order to allow rule matching modulo `Nat` equations. To get reasonable performance with large simpsets, we'll have to figure out how to do this another way.

**Comment by brianhuffman** [comment]
Doing profiling on the `ecdsa.saw` proof reveals that the rewriter is still considering *way* too many potential matches. Across 56 calls to `rewriteSharedTerm`, local function `rewriteTop` is called 239903 times, meaning that it visits nearly a quarter million term AST nodes. But then `scMatch` is called more than *56 million* times (56338787 to be precise), an average of over 200 matches per term node. This is *far* too many.

As it turns out, the problem is the following type of rules from `cryptol_ss`, of which there are many:
```
* (Cryptol.PLogicVec n a pa).not = Prelude.map a a pa.not n
* (Cryptol.PLogicVec n a pa).or = Prelude.zipWith a a a pa.or n
* (Cryptol.PLogicVec n a pa).xor = Prelude.zipWith a a a pa.xor n
* (Cryptol.PLogicWord n).and = Prelude.bvAnd n
* (Cryptol.PLogicWord n).logicZero = Prelude.bvNat n 0
* (Cryptol.PLogicWord n).not = Prelude.bvNot n
* (Cryptol.PLogicWord n).or = Prelude.bvOr n
* (Cryptol.PLogicWord n).xor = Prelude.bvXor n
* (Cryptol.PRingFloat e p).add x y = Prelude.error
                                       (Cryptol.TCFloat e p)
                                       "Unimplemented: (+) Float"
* (Cryptol.PRingFloat e p).int i = Prelude.error (Cryptol.TCFloat e p)
                                     "Unimplemented: toInteger Float"
* (Cryptol.PRingFloat e p).mul x y = Prelude.error
                                       (Cryptol.TCFloat e p)
                                       "Unimplemented: (*) Float"
* (Cryptol.PRingFloat e p).neg x = Prelude.error (Cryptol.TCFloat e p)
                                     "Unimplemented: neg Float"
* (Cryptol.PRingFloat e p).ringZero = Cryptol.PZeroFloat e p
* (Cryptol.PRingFloat e p).sub x y = Prelude.error
                                       (Cryptol.TCFloat e p)
                                       "Unimplemented: (-) Float"
* (Cryptol.PRingFun a b pb).add = Cryptol.funBinary a b pb.add
* (Cryptol.PRingFun a b pb).int i _ = pb.int i
* (Cryptol.PRingFun a b pb).mul = Cryptol.funBinary a b pb.mul
* (Cryptol.PRingFun a b pb).neg = Cryptol.compose a b b pb.neg
* (Cryptol.PRingFun a b pb).ringZero = Cryptol.PZeroFun a b pb.ringZero
* (Cryptol.PRingFun a b pb).sub = Cryptol.funBinary a b pb.sub
* (Cryptol.PRingIntMod n).add = Prelude.intModAdd n
* (Cryptol.PRingIntMod n).int = Prelude.toIntMod n
* (Cryptol.PRingIntMod n).mul = Prelude.intModMul n
* (Cryptol.PRingIntMod n).neg = Prelude.intModNeg n
* (Cryptol.PRingIntMod n).ringZero = Cryptol.PZeroIntMod n
* (Cryptol.PRingIntMod n).sub = Prelude.intModSub n
```
All of these have a left-hand side which is a record field projection of some other term. The problem is that all such rules get inserted into the term net with a variable/wildcard pattern, because we still have the catch-all case in the `termPat`/`termFPat` that is a direct descendant of the one mentioned in my earlier comment from 2022.
https://github.com/GaloisInc/saw-script/blob/ce68251d768d68161a055c6856f42efd17b28dbc/saw-core/src/SAWCore/Conversion.hs#L111-L120

This still needs to get fixed, in the way I suggested 3 years ago. It never makes sense to insert a rewrite rule into a simpset that has a plain wildcard pattern, so we should probably also add a check for this.

Fixing this bug will probably speed up the ecdsa proof significantly.
